impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,198 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
-                                             .getMethod("getUserPrincipal")
-                                             .invoke(
-                                                 Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
-                                                      .getMethod("getTomcatPrincipal")
-                                                      .invoke(principal))
-                            
-                            ;
-                        } catch (Exception e) {
-                            
-                        }
-                    break;
-            }
+        // Check for Servlet
+        try {
+            return CDI.current().select(HttpServletRequest.class).get().getUserPrincipal();
+        } catch (Exception e) {
+            // Not inside an HttpServletRequest
+        }
+
+        // Check for EJB
+        EJBContext ejbContext = EJB.getEJBContext();
+        if (ejbContext != null) {
+            // The EJB returned value must be verified for its "unauthenticated name" since it's vendor specific
+            return getVendorCallerPrincipal(ejbContext.getCallerPrincipal(), true);
+        }
 
+        for (Principal principal : principals) {
+            // If the principal is a standard one, assume it's the "application" caller principal and return it
             if (CallerPrincipal.class.isAssignableFrom(principal.getClass())) {
                 return principal;
             }
-            
-            try {
-                return CDI.current().select(HttpServletRequest.class).get().getUserPrincipal();
-            } catch (Exception e) {
-                // Not inside an HttpServletRequest
-            }
-         
-            EJBContext ejbContext = EJB.getEJBContext();
-            if (ejbContext != null) {
-                String principalName = ejbContext.getCallerPrincipal().getName();
-             
-                // EJBs return empty/null name for unauthenticated callers
-                return principalName != null && !principalName.isEmpty() ? ejbContext.getCallerPrincipal() : null;
+
+            // Do some checks to determine it from vendor specific data
+            Principal vendorCallerPrincipal = getVendorCallerPrincipal(principal, false);
+            if (vendorCallerPrincipal != null) {
+                return vendorCallerPrincipal;
             }
+        }
 
-            return null;
+        return null;
+    }
+
+    /**
+     * Get the underlying caller principal based on vendor specific (e.g.: class
+     * names, EJB unauthenticated name, etc)
+     *
+     * @param principal
+     * @return
+     */
+    private Principal getVendorCallerPrincipal(Principal principal, boolean isEjb) {
+        switch (principal.getClass().getName()) {
+            case "org.glassfish.security.common.PrincipalImpl": // GlassFish/Payara
+                return isEjb ? getAuthenticatedPrincipal(principal, "ANONYMOUS") : principal;
```

Comment:
Instead of writing `isEjb ? getAuthenticatedPrincipal(principal, "ANONYMOUS") : principal` on each line here, the `isEjb` flag could be passed to `getAuthenticatedPrincipal()`, so the decision logic would be implemented just once, rather than repeated for each principal type.